### DISCUSSION TITLE
Best practice to retrieve a Content Type in a Controller using the technical name or his alias

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#15336] [Q&A] [answered]
Hi,

I'm implementing my own theme module and attempting to retrieve the "Content Type" I created in the admin backoffice, from within the controller. I need to do it when I get the page, in a GET action that serves my Razor view.

I tried using `ISession` to retrieve all the Content Types with the specific name "Index" (the technical name I chose), but I believe this approach is not ideal, because **a query like this one can be really expensive**. A potentially better method might involve using the "alias" name. Of course I'm calling ISession in my controller just for this test, in a real environment I will implement an handler.

```
  public class HomeController : Controller
  {
      private readonly IContentManager _contentManager;
      private readonly YesSql.ISession _session;

      public HomeController(YesSql.ISession session, IContentManager contentManager)
      {
          _session = session;
          _contentManager = contentManager;
      }

      public async Task<IActionResult> Index()
      {
          var myContentItem = await _session.Query<ContentItem, ContentItemIndex>(ci => ci.ContentType == "Index").FirstOrDefaultAsync();

          return View();
      }
  }
```

Upon reviewing the documentation, I saw that there is a [tag helper](https://docs.orchardcore.net/en/main/docs/reference/modules/Contents/#razor-helper) for the frontend that facilitates this process in Razor views. But **my goal is to perform this action inside the controller** (later in an handler class) **and then map the Content Item to a class. but I want to do this inside the controller and then map the `ContentItem` in a class.** 

According to what I have found, `ContentManager` doesn't implement a `GetAsync` [method to retrieve a Content Item](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.ContentManagement.Abstractions/IContentManager.cs) by technical name, neither by alias name. 

How can I do this? **Is there documentation that explains the best practice for retrieving a specific Content Item inside a class rather than in a view?** Should I use the APIs?

Thank you

**Comment by MarGraz** [comment, marked as answer]
Hi everyone,

I read here "[Refactor IContentAliasManager to IContentHandleManager](https://orcharddojo.net/blog/pager-with-page-size-refactor-icontentaliasmanager-to-icontenthandlemanager-this-week-in-orchard-25-09-2020)" that the `IContentAliasManager` was refactored, and now it's been in use `IContentHandleManager`, that expose one `GetContentItemIdAsync(string handle)` method trhough it is possible to retrieve a Content Item **by Id**: `"contentitemid:{the content idem id string}"`,  **by slug**: `"slug:{slug name string}"`, and **by alias**: `"alias:{alias name string}"`. 

**I was interested to retrieve my Content Item by using the alias**, and after adding the "Alias" _content parts_, in my "Index" _content item_, it was possible to retrieve the Id using the alias, and then search the specific "Index" content item. Here an implementation example (_of course I will do this not in the controller, later I will implement it in an handler_):

```
  public class HomeController : Controller
  {
      private readonly IContentHandleManager _contentHandleManager;
      private readonly IContentManager _contentManager;      

      public HomeController(IContentHandleManager contentHandleManager, IContentManager contentManager)
      {
          _contentHandleManager = contentHandleManager;
          _contentManager = contentManager;
      }

      public async Task<IActionResult> Index()
      {
          string contentItemId = await _contentHandleManager.GetContentItemIdAsync("alias:my-test-item");

          ContentItem myContentItem = await _contentManager.GetAsync(contentItemId);

          return View();
      }
  }
```

If you are curious to see how the implementation of `IContentHandleManager` works under the hood, here you can find the three concrete implementations for the [alias](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Alias/Services/AliasPartContentHandleProvider.cs), here the one for the [slug](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.Autoroute.Core/Services/AutorouteHandleProvider.cs), and here the one for the [content item id](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Contents/Services/ContentItemIdHandleProvider.cs).

Thank you

**Reply by MarGraz** [reply]
@Piedone, I have a question for you. 

I was unable to find this topic in the documentation. If what I have done is correct, could I add this to the docs? If yes, could you show me which MD file I should add this to?

Thank you

**Reply by Piedone** [reply]
Awesome!

If you can make it into a complete story/tutorial (i.e. have a need as a starting point, and arrive at a solution), then it would be good as a CMS Guide under https://docs.orchardcore.net/en/latest/docs/guides/. See https://docs.orchardcore.net/en/latest/docs/topics/docs-contributions/.

**Reply by MarGraz** [reply]
@Piedone I'm writing a quick startup guide and for sure I will add this topic in the guide.

Thank you
